### PR TITLE
Bring memaccounting dump memory usage back

### PR DESCRIPTION
**Pull request by baobao0206**
Add gp_dump_memory_usage guc back, and add MemoryAccounting_SaveToFile
back to MemoryAccounting_Reset.
Use AllocateFile and FreeFile to replace fopen and fclose. It could
avoid leaking the file descriptor on error.

We already have unittest to cover the code.
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`

### CODE REVIEW

**Comment by hlinnaka** [comment]
`memory_profiler_run_id`, `memory_profiler_dataset_id`, and `memory_profiler_query_id` are used to form the filename in `MemoryAccounting_SaveToFile()`, without proper escaping. Since they are PGC_USERSET, that allows the user to overwrite arbitrary OS files.

SaveMemoryBufToDisk() should use AllocateFile() instead of fopen(), to avoid leaking the file descriptor on error

The tests seem unrelated to the new GUC. Is that correct?

I don't understand the tests. There are no comments. The "oom_sleep" test is completely inside a start/end_ignore block, which seems pointless. Is there any difference between the *_optimizer.out outputs and the regular ones?

**Comment by baobao0206** [comment]
> `memory_profiler_run_id`, `memory_profiler_dataset_id`, and `memory_profiler_query_id` are used to form the filename in `MemoryAccounting_SaveToFile()`, without proper escaping. Since they are PGC_USERSET, that allows the user to overwrite arbitrary OS files.
> 
> SaveMemoryBufToDisk() should use AllocateFile() instead of fopen(), to avoid leaking the file descriptor on error
> 
> The tests seem unrelated to the new GUC. Is that correct?
> 
> I don't understand the tests. There are no comments. The "oom_sleep" test is completely inside a start/end_ignore block, which seems pointless. Is there any difference between the *_optimizer.out outputs and the regular ones?

@hlinnaka The location of the file is in the pg_log directory, and prefix by 'mem_'. So they should not overwrite system files. 
Have replace fopen and fclose with AllocateFile and FreeFile.
The tests are related to other gucs related to memaccounting. And they are moved from tinc test to isolation test. So, I make another pr to enable the tests.

**Comment by hlinnaka** [comment]
> > `memory_profiler_run_id`, `memory_profiler_dataset_id`, and `memory_profiler_query_id` are used to form the filename in `MemoryAccounting_SaveToFile()`, without proper escaping. Since they are PGC_USERSET, that allows the user to overwrite arbitrary OS files.
> > SaveMemoryBufToDisk() should use AllocateFile() instead of fopen(), to avoid leaking the file descriptor on error
> > The tests seem unrelated to the new GUC. Is that correct?
> > I don't understand the tests. There are no comments. The "oom_sleep" test is completely inside a start/end_ignore block, which seems pointless. Is there any difference between the *_optimizer.out outputs and the regular ones?
> 
> @hlinnaka The location of the file is in the pg_log directory, and prefix by 'mem_'. So they should not overwrite system files.
> Have replace fopen and fclose with AllocateFile and FreeFile.
> The tests are related to other gucs related to memaccounting. And they are moved from tinc test to isolation test. So, I make another pr to enable the tests.

There's no checking or `/` or `..` in the identifiers. Yeah, maybe you're safe in practice, because there can't be any suitable subdirectories in pg_log that you could latch on to, but it sure screams "directory traversal!" to me.

**Comment by berlin-ab** [comment]
What's the status of this PR?

**Comment by Eulerizeit** [comment]
@hardikar could you take a look at this please?

File: src/backend/utils/misc/guc_gp.c [hunk 1]
```diff
@@ -166,6 +166,7 @@ char	   *memory_profiler_run_id = "none";
 char	   *memory_profiler_dataset_id = "none";
 char	   *memory_profiler_query_id = "none";
 int			memory_profiler_dataset_size = 0;
+bool		gp_dump_memory_usage = FALSE;
```
Comment (hardikar) [inline src/backend/utils/misc/guc_gp.c]: s/FALSE/false/?

**Comment by berlin-ab** [comment]
The TINC tests came as a part of another PR, right @baobao0206 . Let's get this PR in.

**Comment by l-wang** [comment]
There are already some unit test on it. We are thinking wether some regression tests are needed.

**Comment by berlin-ab** [comment]
@baobao0206 This looks approved, but it's been a while since it's been touched. Would you please take a quick look and get it in if we can?

**Comment by berlin-ab** [comment]
@baobao0206 Do you plan on continuing this PR, or should we close it?

**Comment by hlinnaka** [comment]
So, are we not bringing the memory accounting stuff back? Can we remove it all? I'd love to remove it if it's not used, because it's causing merge conflicts as we merge with upstream, and also making palloc/pfree() noticeably slower.

**Comment by ashwinstar** [comment]
re-opening to not loose track till @hlinnaka's question is answered.

**Comment by baobao0206** [comment]
Close it  for we already have new plans on it

**Pull request closed without merging**